src/ExpressionEvaluator/CSharp/Source/ExpressionCompiler/CompilationContext.cs

```diff
@@ -828,14 +830,35 @@ private static int IndexOfMatchingAssembly(AssemblyIdentity referenceIdentity, I
                     binder);
             }
 
-            binder = new EEMethodBinder(method, substitutedSourceMethod, binder);
+            Binder originalRootBinder = null;
+            CSharpSyntaxNode declaredLocalsScopeDisagnator = null;
+            var executableBinder = new ExecutableCodeBinder(syntax, substitutedSourceMethod, binder,
+                                              (rootBinder, declaredLocalsScopeDisagnatorOpt) =>
+                                              {
+                                                  originalRootBinder = rootBinder;
+                                                  declaredLocalsScopeDisagnator = declaredLocalsScopeDisagnatorOpt;
+                                                  binder = new EEMethodBinder(method, substitutedSourceMethod, rootBinder);
 
-            if (methodNotType)
+                                                  if (methodNotType)
+                                                  {
+                                                      binder = new SimpleLocalScopeBinder(method.LocalsForBinding, binder);
+                                                  }
+
+                                                  return binder;
+                                              });
+
+            // We just need to trigger the process of building the binder map
+            // so that the lambda above was executed.
+            executableBinder.GetBinder(syntax);
+
+            Debug.Assert(originalRootBinder != null);
+            Debug.Assert(executableBinder.Next != binder);
+
+            if (declaredLocalsScopeDisagnator != null)
             {
-                binder = new SimpleLocalScopeBinder(method.LocalsForBinding, binder);
+                declaredLocals = originalRootBinder.GetDeclaredLocalsForScope(declaredLocalsScopeDisagnator);
             }
```

Comment:
It should be set, looks like a bug in the compiler, it doesn't require out parameter to be fully initialized.